pkg/ovsdb/transact.go

```diff
@@ -0,0 +1,1094 @@
+	return nil
+}
+
+/* insert */
+func preInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	if ovsOp.UUIDName == "" {
+		return nil
+	}
+	etcdGetTable(txn, ovsOp.Table)
+	return nil
+}
+
+func doInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	if ovsOp.UUIDName != "" {
+		for _, row := range txn.data.Table(ovsOp.Table) {
+			if (*row)[COL_UUIDNAME] == ovsOp.UUIDName {
+				err := errors.New(E_DUP_UUIDNAME)
+				ovsResult.Error = err.Error()
+				return err
+			}
+		}
+	}
+	ovsResult.Count = ovsResult.Count + 1
+	uuid := guuid.New().String() /* generate RFC4122 UUID */
+	row := txn.data.Row(ovsOp.Table, uuid)
+	*row = ovsOp.Row // supporting read-afrer-write
+	return etcdPutRow(txn, ovsOp.Table, uuid, row)
+}
+
+/* select */
+func preSelect(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return etcdGetByWhere(txn, ovsOp, ovsResult)
+}
+
+func doSelect(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	for _, row := range txn.data.Table(ovsOp.Table) {
+		ok, err := isRowSelectedByWhere(row, ovsOp.Where)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		resultRow := reduceRowByColumns(row, ovsOp.Columns)
+		ovsResult.Count = ovsResult.Count + 1
+		ovsResult.Rows = append(ovsResult.Rows, *resultRow)
+	}
+	return nil
+}
+
+/* update */
+func preUpdate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return etcdGetByWhere(txn, ovsOp, ovsResult)
+}
+
+func doUpdate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	for uuid, row := range txn.data.Table(ovsOp.Table) {
+		ok, err := isRowSelectedByWhere(row, ovsOp.Where)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		ovsResult.Count = ovsResult.Count + 1
+		err = RowUpdate(row, ovsOp.Row)
+		if err != nil {
+			return err
+		}
+		*(txn.data.Row(ovsOp.Table, uuid)) = *row
+		etcdPutRow(txn, ovsOp.Table, uuid, row)
+	}
+	return nil
+}
+
+/* mutate */
+func preMutate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return etcdGetByWhere(txn, ovsOp, ovsResult)
+}
+
+func doMutate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	for uuid, row := range txn.data.Table(ovsOp.Table) {
+		ok, err := isRowSelectedByWhere(row, ovsOp.Where)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		ovsResult.Count = ovsResult.Count + 1
+		err = RowMutate(row, ovsOp.Mutations)
+		if err != nil {
+			return err
+		}
+		*(txn.data.Row(ovsOp.Table, uuid)) = *row
+		etcdPutRow(txn, ovsOp.Table, uuid, row)
+	}
+	return nil
+}
+
+/* delete */
+func preDelete(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return etcdGetByWhere(txn, ovsOp, ovsResult)
```

Comment:
if you have uuid, you don't need fetch

Reply:
as explain, creating different paths of execution (with added code complexity, testing, etc.) needs justification. the general justification of "better performance" is counter-productive as it will force optimized code (and complexity) in places these are not needed. Currently focusing on correctness and simplicity it think its wrong to bring-in optimizations here (and elsewhere in the code) and this preliminary stage.